numerics/newhall_body.hpp

```diff
@@ -96,187 +96,43 @@ struct NewhallAppromixator {
-struct NewhallAppromixator<Vector, 12, Evaluator> {
-  static FixedVector<Vector, 13> HomogeneousCoefficients(
-      FixedVector<Vector, 2 * divisions + 2> const& qv,
-      Vector& error_estimate) {
-    error_estimate =
-        newhall_c_matrix_чебышёв_degree_12_divisions_8_w04.row<12>() * qv;
-    return newhall_c_matrix_monomial_degree_12_divisions_8_w04 * qv;
-  }
-};
-
-template<typename Vector,
-         template<typename, typename, int> class Evaluator>
-struct NewhallAppromixator<Vector, 13, Evaluator> {
-  static FixedVector<Vector, 14> HomogeneousCoefficients(
-      FixedVector<Vector, 2 * divisions + 2> const& qv,
-      Vector& error_estimate) {
-    error_estimate =
-        newhall_c_matrix_чебышёв_degree_13_divisions_8_w04.row<13>() * qv;
-    return newhall_c_matrix_monomial_degree_13_divisions_8_w04 * qv;
-  }
-};
-
-template<typename Vector,
-         template<typename, typename, int> class Evaluator>
-struct NewhallAppromixator<Vector, 14, Evaluator> {
-  static FixedVector<Vector, 15> HomogeneousCoefficients(
-      FixedVector<Vector, 2 * divisions + 2> const& qv,
-      Vector& error_estimate) {
-    error_estimate =
-        newhall_c_matrix_чебышёв_degree_14_divisions_8_w04.row<14>() * qv;
-    return newhall_c_matrix_monomial_degree_14_divisions_8_w04 * qv;
-  }
-};
-
-template<typename Vector,
-         template<typename, typename, int> class Evaluator>
-struct NewhallAppromixator<Vector, 15, Evaluator> {
-  static FixedVector<Vector, 16> HomogeneousCoefficients(
-      FixedVector<Vector, 2 * divisions + 2> const& qv,
-      Vector& error_estimate) {
-    error_estimate =
-        newhall_c_matrix_чебышёв_degree_15_divisions_8_w04.row<15>() * qv;
-    return newhall_c_matrix_monomial_degree_15_divisions_8_w04 * qv;
-  }
-};
-
-template<typename Vector,
-         template<typename, typename, int> class Evaluator>
-struct NewhallAppromixator<Vector, 16, Evaluator> {
-  static FixedVector<Vector, 17> HomogeneousCoefficients(
-      FixedVector<Vector, 2 * divisions + 2> const& qv,
-      Vector& error_estimate) {
-    error_estimate =
-        newhall_c_matrix_чебышёв_degree_16_divisions_8_w04.row<16>() * qv;
-    return newhall_c_matrix_monomial_degree_16_divisions_8_w04 * qv;
+#define PRINCIPIA_NEWHALL_APPROXIMATOR_SPECIALIZATION(degree)                  \
+  template<typename Vector,                                                    \
+           template<typename, typename, int> class Evaluator>                  \
+  struct NewhallAppromixator<Vector, (degree), Evaluator> {                    \
+    static FixedVector<Vector, ((degree) + 1)> HomogeneousCoefficients(        \
+        FixedVector<Vector, 2 * divisions + 2> const& qv,                      \
+        Vector& error_estimate) {                                              \
+      error_estimate =                                                         \
+          newhall_c_matrix_чебышёв_degree_##degree##_divisions_8_w04.row<      \
```

Comment:
I wonder whether we could generate std::arrays or tuples or something to obviate the need for these specializations...

Reply:
Let's not generate code when we have a perfectly good solution in the language.

Reply:
No, I mean since we generate newhall_c_matrix_чебышёв_degree_7_divisions_8_w04 etc. already, perhaps we could generate those as a tuple so that the token joining here becomes unnecessary and it can just be done via template magic. We still need the macros for the cases, where the degree is not a constant expression.